### ISSUE TITLE
[QUESTION]-Why BulkApi v1 Parallel Mode does NOT create batches asynchronously ?

### ISSUE THREAD

**Issue by grand-wazoo** [forcedotcom/SFDX-Data-Move-Utility#423]
**Describe the bug**
When upserting or deleting data in target org. sfdmu doesn't create batches at the same time. Instead, a new batch is created when previous batch completes. Both the log and salesforce Bulk Data Load Jobs details indicate that. As a result all data is upserted in serial mode and sfdmu doesn't take advantage of Parallel execution. See screenshots from Bulk Data Load Jobs details:

![job_Detail](https://user-images.githubusercontent.com/86124200/152693207-2963b859-2921-4535-9ecb-51c289e3bcfe.png)
![batch_Detail](https://user-images.githubusercontent.com/86124200/152693214-5a8836b5-1561-4110-9ff0-e42d585bc3ef.png)

**To Reproduce**
Push a lot of data using BulkApi v1 (>> than the bulkApiV1BatchSize). Notice how batches are created sequentially.

**Expected behavior**
All batches within a job should be create near simultaneously

**export.json**

> {
>   "objects": [
>     {
>       "query": "Select Id  FROM TimeSlot Where CHS_Source_Id__c != ''",
>       "operation": "Delete",
>       "externalId": "CHS_Source_Id__c"
>     },
>     {
>       "query": "Select Id  FROM OperatingHours Where CHS_Source_Id__c != ''",
>       "operation": "Delete",
>       "externalId": "CHS_Source_Id__c"
>     }
>   ],
>   "promptOnMissingParentObjects": false,
>   "promptOnIssuesInCSVFiles": false,
>   "bulkApiVersion": "1.0",
>   "bulkApiV1BatchSize": 5000,
>   "keepObjectOrderWhileExecute": true,
>   "concurrencyMode": "Parallel"
> }

**Log file**
[TimeSlot_log.txt](https://github.com/forcedotcom/SFDX-Data-Move-Utility/files/8010734/TimeSlot_log.txt)

**Comment by hknokh** [comment]
Hello.
By the implementation and for some techincal reasons for record **deletion** the Tool always use serial mode even you set concurrencyMode = "Parallel".
For udates/inserts it should work in parallel. Please recheck and report if it's not so.

**Comment by grand-wazoo** [comment]
Upsert acts the exact same way:

![job_Detail_upsert](https://user-images.githubusercontent.com/86124200/153522597-0458235a-9a17-4326-91da-76361a878d72.png)

![batch_Detail_Upsert](https://user-images.githubusercontent.com/86124200/153522621-5c1244aa-f7db-4e10-ba34-b85ca6f440db.png)

[TimeSlot_log_insert.txt](https://github.com/forcedotcom/SFDX-Data-Move-Utility/files/8045327/TimeSlot_log_insert.txt)


> {
>   "objects": [
>     {
>       "query": "Select Id, CHS_Source_Id__c, Account__c, Name, TimeZone, Type__c FROM OperatingHours  Where CHS_Source_Id__c !=''",
>       "operation": "Upsert",
>       "externalId": "CHS_Source_Id__c"
>     },
>     {
>       "query": "Select Id,CHS_Source_Id__c,DayOfWeek,EndTime,OperatingHoursId,StartTime,Type,X24_Hours__c FROM TimeSlot  Where CHS_Source_Id__c !=''",
>       "operation": "Upsert",
>       "externalId": "CHS_Source_Id__c"
>     }
>   ],
>   "promptOnMissingParentObjects": false,
>   "promptOnIssuesInCSVFiles": false,
>   "bulkApiVersion": "1.0",
>   "bulkApiV1BatchSize": 5000,
>   "concurrencyMode": "Parallel"
> }

This is the command I am executing:

sfdx sfdmu:run --sourceusername csvfile --targetusername $AgentName --path $workingdir --verbose --canmodify --noprompt

**Comment by hknokh** [comment]
All is correct with the SFDMU as per your description.

- First of all in your case all what you see is in your log is **the same bulk job** but a set of batches **under the same job.** Each batch has size of  5000 records as u set bulkApiV1BatchSize = 5000 in your config.
Pay attention below that the **job ID id is the same** for every batch:
![image](https://user-images.githubusercontent.com/15181301/153549142-097dbf8c-df4c-4ea9-b7b9-fc56dda675e2.png)

- Second, the parallel mode means that the bulk job will be created using concurencyMode = Parallel meaning that the batches under the same job will be executed by the SF (internally) in parallel.

In summary I see that the SFDMU creates job in parallel concurency mode as expected and splits the records into batches of 5000 records and sends these baches to the SF for processing at once.

All is as expected.

For your ref, I found this article explains well the concurrency mode:
https://wdcigroup.net/salesforce-integration-tips-using-the-bulk-api-serial-or-parallel-mode-options/

**Comment by grand-wazoo** [comment]
I asked specifically about the batches, not the jobs. sfmdu does NOT send the batches at once. Both the log and the screenshot of the batch start times confirms it:
1st batch created @ 15:33:20. Completed 15:33:39
2nd batch created @ 15:33:40. Completed 15:33:57
3rd batch created @ 15:33:58. Completed 15:34:15

TimeSlot gets upserted fairly quickly, so the issue is not as apparent. I've only been using it as an example, because it is simple, but other objects are slower and an upsert of large dataset takes hours, because it is executes 5K records at a time and waits for  the batch to complete, before creating a new one.

**Comment by hknokh** [comment]
Each batch takes a time to be sent to the salesforce server. 
SFDMU IS NOT SENDING TO SF IN PARALLEL THREADS!
The SF executes in parallel what the SFDMU sent  sequentally.
This is the current implemetation.
